### PR TITLE
UCM/EVENT: Replace mutex by read-write lock.

### PR DESCRIPTION
**Pull request by yosefe**

### CODE REVIEW

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/3232/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr__NEW/771/ for details.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/2709/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/2712/ for details (Mellanox internal link).

**Comment by yosefe** [comment]
bot:retest

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/3245/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr__NEW/779/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/2720/ for details (Mellanox internal link).

**Comment by yosefe** [comment]
@alex-mikheev @brminich pls take a look

File: src/ucm/event/event.c [hunk 1]
```diff
@@ -123,23 +122,12 @@ static void ucm_event_dispatch(ucm_event_type_t event_type, ucm_event_t *event)
...
-    self = pthread_self();
-    if (pthread_mutex_trylock(&ucm_event_lock) != 0) {
-        ucs_memory_cpu_load_fence();
-        if (ucm_event_lock_owner == self) {
-            ucm_fatal("Detected recursive call to memory event handler");
-        }
-        pthread_mutex_lock(&ucm_event_lock);
-    }
-    ucm_event_lock_owner = self;
+    pthread_rwlock_rdlock(&ucm_event_lock);
```
Comment (brminich) [inline src/ucm/event/event.c]: Should not we check for return value? (to make sure we do not own the lock already, also max num of simultaneous read lock may be exceeded)

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/3254/ for details.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/2727/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr__NEW/786/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/3257/ for details.

**Comment by yosefe** [comment]
bot:retest

File: src/ucm/event/event.c [hunk 2]
```diff
@@ -123,23 +123,23 @@ static void ucm_event_dispatch(ucm_event_type_t event_type, ucm_event_t *event)
...
-        ucs_memory_cpu_load_fence();
-        if (ucm_event_lock_owner == self) {
-            ucm_fatal("Detected recursive call to memory event handler");
-        }
-        pthread_mutex_lock(&ucm_event_lock);
+static void ucm_event_enter_exclusive()
+{
+    int ret = pthread_rwlock_wrlock(&ucm_event_lock);
+    if (ret != 0) {
+        ucm_fatal("pthread_rwlock_wrlock() failed: %s", strerror(ret));
```
Comment (alex-mikheev) [inline src/ucm/event/event.c]: imho assert is enough here

File: src/ucm/event/event.c [hunk 3]
```diff
@@ -123,23 +123,23 @@ static void ucm_event_dispatch(ucm_event_type_t event_type, ucm_event_t *event)
 
 static void ucm_event_enter()
 {
-    pthread_t self;
+    int ret = pthread_rwlock_rdlock(&ucm_event_lock);
+    if (ret != 0) {
+        ucm_fatal("pthread_rwlock_rdlock() failed: %s", strerror(ret));
```
Comment (alex-mikheev) [inline src/ucm/event/event.c]: can busy loop on EAGAIN

File: src/ucm/event/event.c (hunk 2 above)
Comment (yosefe) [inline src/ucm/event/event.c]: we don't have assert in ucm (which is standalone and does not use ucs)

File: src/ucm/event/event.c (hunk 3 above)
Comment (yosefe) [inline src/ucm/event/event.c]: fixed

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/3262/ for details.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr__NEW/796/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/3264/ for details.

**Comment by yosefe** [comment]
bot:mlx:retest

**Comment by yosefe** [comment]
bot:mlx:retest

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr__NEW/798/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/2746/ for details (Mellanox internal link).

**Comment by yosefe** [comment]
@alex-mikheev pls take a look again

**Pull request merged**